src/connection.rs

```diff
@@ -108,10 +109,18 @@ impl Connection {
                 .await
                 .map(|c| Connection::Unix(BufReader::new(BufWriter::new(c))))
                 .map_err(Error::Connect),
-            Addr::Tcp(url) | Addr::Unknown(url) => TcpStream::connect(url)
-                .await
-                .map(|c| Connection::Tcp(BufReader::new(BufWriter::new(c))))
-                .map_err(Error::Connect),
+            Addr::Tcp(url) | Addr::Unknown(url) => {
+                let addr: SocketAddr = url.parse().map_err(|e| {
+                    Error::Connect(io::Error::new(
+                        io::ErrorKind::InvalidInput,
+                        format!("Invalid address: {}", e),
+                    ))
+                })?;
+                let socket = TcpSocket::new_v4().map_err(Error::Connect)?;
+                socket.set_nodelay(true).map_err(Error::Connect)?;
```

Comment:
I think this could be an option we passed into the Connection.

Reply:
IMO it should just be the default behaviour; [Go disables it by default](https://github.com/golang/go/blob/master/src/net/tcpsock.go#L259-L271). Here's a snippet from the Wikipedia page of where it _is_ useful.

> The RFC describes what Nagle calls the "small-packet problem", where an application repeatedly emits data in small chunks, frequently only 1 [byte](https://en.wikipedia.org/wiki/Byte) in size. Since [TCP](https://en.wikipedia.org/wiki/Transmission_Control_Protocol) packets have a 40-byte header (20 bytes for TCP, 20 bytes for [IPv4](https://en.wikipedia.org/wiki/IPv4)), this results in a 41-byte packet for 1 byte of useful information, a huge overhead. This situation often occurs in [Telnet](https://en.wikipedia.org/wiki/Telnet) sessions, where most keypresses generate a single byte of data that is transmitted immediately. Worse, over slow links, many such packets can be in transit at the same time, potentially leading to [congestion collapse](https://en.wikipedia.org/wiki/Congestion_collapse).

https://en.wikipedia.org/wiki/Nagle%27s_algorithm

Reply:
Fun read https://brooker.co.za/blog/2024/05/09/nagle.html